R/resolve_phylopic.R

```diff
@@ -0,0 +1,233 @@
+    url <- paste0("https://www.marinespecies.org/rest/",
+                  "AphiaRecordsByMatchNames?marine_only=false&",
+                  "scientificnames%5B%5D=", name)
+    res <- GET(url = url)
+    if (length(content(res)) == 0) stop("No results returned from the API.")
+    jsn <- response_to_JSON(res)
+    ids <- jsn[[1]]$AphiaID[1]
+    name_vec <- jsn[[1]]$scientificname[1]
+    if (hierarchy) {
+      url <- paste0("https://www.marinespecies.org/rest/",
+                    "AphiaClassificationByAphiaID/", ids)
+      res <- GET(url = url)
+      jsn <- response_to_JSON(res)
+      lst_sub <- jsn
+      ids <- character()
+      name_vec <- character()
+      while ("child" %in% names(lst_sub)) {
+        ids <- c(ids, lst_sub$AphiaID)
+        name_vec <- c(name_vec, lst_sub$scientificname)
+        lst_sub <- lst_sub$child
+      }
+      ids <- rev(ids)
+      name_vec <- rev(name_vec)
+    }
+  } else if (api == "paleobiodb.org") {
+    # check api is online
+    headers <- curlGetHeaders("https://paleobiodb.org/data1.2/")
+    if (attr(headers, "status") != 200) {
+      stop("Paleobiology Database is not available or you have no internet
+           connection.")
+    }
+    namespace <- "txn"
+    url <- paste0("https://paleobiodb.org/data1.2/taxa/auto.json?",
+                  "limit=1&name=", name)
+    res <- GET(url = url)
+    jsn <- response_to_JSON(res)
+    if ("errors" %in% jsn ||
+        length(jsn$records) == 0) stop("No results returned from the API.")
+    ids <- jsn$records$oid[1]
+    name_vec <- jsn$records$nam[1]
+    if (hierarchy) {
+      url <- paste0("https://paleobiodb.org/data1.2/taxa/list.json?",
+                    "rel=all_parents&", "id=txn:", ids)
+      res <- GET(url = url)
+      jsn <- response_to_JSON(res)
+      ids <- rev(gsub("txn:", "", jsn$records$oid))
+      name_vec <- rev(jsn$records$nam)
+    }
+  } else if (api == "opentreeoflife.org") {
+    # check api is online
+    headers <- curlGetHeaders("https://api.opentreeoflife.org/")
+    if (attr(headers, "status") != 200) {
+      stop("Open Tree of Life is not available or you have no internet
+           connection.")
+    }
+    namespace <- "taxonomy"
+    url <- "https://api.opentreeoflife.org/v3/tnrs/autocomplete_name"
+    res <- POST(url = url, encode = "json",
+                body = list("name" = URLdecode(name)))
+    jsn <- response_to_JSON(res)
+    if (length(jsn) == 0) stop("No results returned from the API.")
+    ids <- jsn$ott_id[1]
+    name_vec <- jsn$unique_name[1]
+    if (hierarchy) {
+      url <- "https://api.opentreeoflife.org/v3/taxonomy/taxon_info"
+      res <- POST(url = url, encode = "json",
+                  body = list("include_lineage" = TRUE, "ott_id" = ids))
+      jsn <- response_to_JSON(res)
+      ids <- c(ids, jsn$lineage$ott_id)
+      name_vec <- c(name_vec, jsn$lineage$unique_name)
+    }
+  } else {
+    stop("Invalid value for `api`. See the documentation for valid options.")
```

Comment:
Might be nice to just list the options here, for convenience. (This function already determines what is and is not valid, anyway.)